src/pe/certificate_table.rs

```diff
@@ -0,0 +1,146 @@
+use crate::error;
+use scroll::Pread;
+
+use alloc::string::ToString;
+use alloc::vec::Vec;
+
+#[repr(u16)]
+#[derive(Debug, PartialEq, Copy, Clone)]
+enum CertificateAttributeRevision {
+    Revision1_0 = 0x0100,
+    Revision2_0 = 0x0200,
+}
+
+impl TryFrom<u16> for CertificateAttributeRevision {
+    type Error = error::Error;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        Ok(match value {
+            x if x == CertificateAttributeRevision::Revision1_0 as u16 => {
+                CertificateAttributeRevision::Revision1_0
+            }
+            x if x == CertificateAttributeRevision::Revision2_0 as u16 => {
+                CertificateAttributeRevision::Revision2_0
+            }
+            _ => {
+                return Err(error::Error::Malformed(
+                    "Invalid certificate attribute revision".to_string(),
+                ))
+            }
+        })
+    }
+}
+
+#[repr(u16)]
+#[derive(Debug)]
+enum CertificateAttributeType {
+    X509 = 0x0001,
+    PkcsSignedData = 0x0002,
+    Reserved1 = 0x0003,
+    TsStackSigned = 0x0004,
+}
+
+impl TryFrom<u16> for CertificateAttributeType {
+    type Error = error::Error;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        Ok(match value {
+            x if x == CertificateAttributeType::X509 as u16 => CertificateAttributeType::X509,
+            x if x == CertificateAttributeType::PkcsSignedData as u16 => {
+                CertificateAttributeType::PkcsSignedData
+            }
+            x if x == CertificateAttributeType::Reserved1 as u16 => {
+                CertificateAttributeType::Reserved1
+            }
+            x if x == CertificateAttributeType::TsStackSigned as u16 => {
+                CertificateAttributeType::TsStackSigned
+            }
+            _ => {
+                return Err(error::Error::Malformed(
+                    "Invalid certificate attribute type".to_string(),
+                ))
+            }
+        })
+    }
+}
+
+#[derive(Clone, Pread)]
+pub struct CertificateAttributeHeader {
+    length: u32,
+    revision: u16,
+    certificate_type: u16,
+}
+
+pub const CERTIFICATE_DATA_OFFSET: u32 = 8;
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct CertificateAttribute<'a> {
+    length: u32,
+    revision: CertificateAttributeRevision,
+    certificate_type: CertificateAttributeType,
+    certificate: &'a [u8],
+}
+
+impl<'a> CertificateAttribute<'a> {
+    pub fn parse(
+        bytes: &'a [u8],
+        current_offset: &mut usize,
+    ) -> Result<CertificateAttribute<'a>, error::Error> {
+        // `current_offset` is moved sizeof(CertificateAttributeHeader) = 8 bytes further.
+        let header: CertificateAttributeHeader = bytes.gread_with(current_offset, scroll::LE)?;
```

Comment:
I could not find a relevant section in the spec about it, but do you know if those fields are specified as little-endian? Or maybe it's arch-dependent, or maybe we don't care since no actually use a bigendian arch?

Reply:
defined as `DWORD` / `WORD`
https://learn.microsoft.com/en-us/windows/win32/api/wintrust/ns-wintrust-win_certificate

which are type aliases for `long`/`short`:
https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types

So I guess it's arch-dependant?

Reply:
```suggestion
        let header: CertificateAttributeHeader = bytes.gread_with(current_offset, scroll::NATIVE)?;
```

https://docs.rs/scroll/latest/scroll/constant.NATIVE.html

Reply:
(`scroll::NATIVE` doesn't cut it, it's host dependent not specific to the target PE)

Reply:
Hmm, do we have a good solution then?

Reply:
From my understanding, all versions of Windows are using little-endian, but the NT kernel on Xbox 360 ran on a big endian (PowerPC) arch.
Can this information be provided by the user, I guess? cc @m4b

Reply:
I'd settle for a todo, I don't think there is a real-world use of bigendian. But I never contributed a commit here, my opinion doesn't count.

Reply:
I think even the Xbox 360 uses little endian in the PE headers, see https://reverseengineering.stackexchange.com/a/17923. All the existing code in this crate for PE uses little endian.

Reply:
Given the existing code in this crate for PE uses little endian, a comment about this would make sense at a project-level or crate-level, so I will pass, except @m4b or @philipc told me otherwise (?).

Reply:
my understanding has always been PE binary records are always little endian, i can't remember where i read this, but i recall it being in the microsoft PE docs